Add support for flash controller with XIP on i.MX RT series

pdgendt opened the pull request:
This change allows writing to the flash while running in XIP mode, and enables `mcuboot` or NVS settings to be used on i.MX RT socs with a single flash chip.
The bootloader application itself should contain the IVT/DCD in the header, but the chainable application doesn't.
The ROM_START_OFFSET defaults to 0x400 otherwise the linker alignment isn't taken into account.

It is required when running with XIP to enable code relocation, because accessing the flash needs to run in RAM.

Tested on i.MX RT1064 internal flash

pdgendt committed:
soc: nxp_imx: Disable IVT/DCD when building for mcuboot

The bootloader application itself should contain the IVT/DCD
in the header, but the chainable application doesn't.

The ROM_START_OFFSET defaults to 0x400 otherwise the linker
alignment isn't taken into account.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

MaureenHelm commented on drivers/flash/CMakeLists.txt:
```diff
@@ -22,6 +22,12 @@ zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI flash_mcux_flexspi.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_NOR flash_mcux_flexspi_nor.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_ESP32 flash_esp32.c)
 
+if(CONFIG_XIP AND CONFIG_FLASH_MCUX_FLEXSPI_NOR AND (CONFIG_CODE_FLEXSPI OR CONFIG_CODE_FLEXSPI2))
```
I think it would be better to add a new `CONFIG_FLASH_MCUX_FLEXSPI_XIP` and move this logic to Kconfig defaults. That would also lend well to moving line 28 to the hal module.

pdgendt replied:
I will change this and make it depend on `CONFIG_CODE_RELOCATION` as well.

Is it safe to assume `ITCM` memory will always be available?

MaureenHelm commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -28,6 +28,17 @@ struct flash_flexspi_data {
+{
+#ifdef CONFIG_XIP
+	/* flexspi bus is labeled FLEXSPI & flexspi2 bus is labeled FLEXSPI1 */
+	return (strcmp(dev->name, "FLEXSPI") == 0 && IS_ENABLED(CONFIG_CODE_FLEXSPI))
+		|| (strcmp(dev->name, "FLEXSPI1") == 0 && IS_ENABLED(CONFIG_CODE_FLEXSPI2));
```
@galak is there a macro in `include/devicetree.h` we could use to replace the strcmp here?

pdgendt replied:
Even for this runtime check? The i.MX series can have 2 `flexspi` busses active where one is running XIP and the other isn't.

pdgendt replied:
I do agree that the string compare isn't ideal especially because it always has to go through the whole name to check if there is a `1` at the end.
How about I compare the device's base address to what is predefined from the DTS?

pdgendt replied:
I did not succeed in creating a compile time `config` member because I couldn't get the macro magic to work, so I've added a `data` member.

If the compile time constant is still preferred, someone will need to give me pointers on how to do this :)

mbolivar-nordic replied:
Using `dev->name` for anything other than debugging is less and less of a good idea as time goes on.

If you have an enumeration of possible devices that `dev` might be and want to find out which one it is, why not store pointers to each struct device and just use pointer `dev == flexspi_dev`, `dev == flexspi2_dev` or so?

pdgendt replied:
@mbolivar-nordic I fully agree, and I've already moved the runtime check to the `init` function where I compare the device's base address [here](https://github.com/zephyrproject-rtos/zephyr/pull/33701/files#diff-76b6fe7650b1c121295403e474946fea492adc9b2565d2dd2f20ccef2b8478eaR124).

I think it could even be optimized to a compile time constant though.

mbolivar-nordic replied:
> @galak is there a macro in `include/devicetree.h` we could use to replace the strcmp here?

Forgot to mention: if you're comparing two known nodes, maybe [DT_SAME_NODE](https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_SAME_NODE)?

pdgendt replied:
@mbolivar-nordic I tried your suggestion like so:

```C
#define MEMC_FLEXSPI_CFG_XIP(node_id) 						\
	COND_CODE_1(DT_SAME_NODE(node_id, DT_NODELABEL(flexspi)), 		\
		(IS_ENABLED(CONFIG_CODE_FLEXSPI)), 				\
		(COND_CODE_1(DT_SAME_NODE(node_id, DT_NODELABEL(flexspi2)), 	\
			(IS_ENABLED(CONFIG_CODE_FLEXSPI2)), (false))))
```

But the `DT_SAME_NODE` does not expand to a literal 1 or 0, wrapping it with `UTIL_BOOL` also didn't work.
This is as far as my macro magic knowledge goes atm :)

mbolivar-nordic replied:
Indeed it does not expand to literal 0 or 1. But neither does comparing the addresses. I will try to find time to take a more detailed look at the PR later but if you have a solution you're happy with that avoids strcmp, never mind me :)

pdgendt replied:
@mbolivar-nordic Your suggestion did get me to a solution if I turn the logic around!

```C
#define MEMC_FLEXSPI_CFG_XIP(node_id) 						\
	COND_CODE_1(CONFIG_CODE_FLEXSPI, 					\
		(DT_SAME_NODE(node_id, DT_NODELABEL(flexspi))), 		\
		(COND_CODE_1(CONFIG_CODE_FLEXSPI2, 				\
			(DT_SAME_NODE(node_id, DT_NODELABEL(flexspi2))), 	\
			(false))))
```

Thanks, will update

MaureenHelm commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -49,53 +52,98 @@ struct flash_flexspi_nor_data {
+				kFLEXSPI_Command_STOP, kFLEXSPI_1PAD, 0),
+	},
+
+	/* Read ID */
+	[NOR_CMD_LUT_SEQ_IDX_READID] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, SPI_NOR_CMD_RDID,
+				kFLEXSPI_Command_READ_SDR, kFLEXSPI_1PAD, 0x04),
+	},
+
+	/* Enable Quad mode */
+	[NOR_CMD_LUT_SEQ_IDX_WRITESTATUSREG] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, SPI_NOR_CMD_WRSR,
+				kFLEXSPI_Command_WRITE_SDR, kFLEXSPI_1PAD, 0x04),
+	},
+
+	/* Enter QPI mode */
+	[NOR_CMD_LUT_SEQ_IDX_ENTERQPI] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, 0x35,
+				kFLEXSPI_Command_STOP, kFLEXSPI_1PAD, 0),
 	},
 
-	[ENTER_QPI] = {
-		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, 0x35,
-				kFLEXSPI_Command_STOP,      kFLEXSPI_1PAD, 0),
+	/* Exit QPI mode */
+	[NOR_CMD_LUT_SEQ_IDX_EXITQPI] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_4PAD, 0xF5,
+				kFLEXSPI_Command_STOP, kFLEXSPI_1PAD, 0),
+	},
+
+	/* Read status register */
+	[NOR_CMD_LUT_SEQ_IDX_READSTATUSREG] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, SPI_NOR_CMD_RDSR,
+				kFLEXSPI_Command_READ_SDR, kFLEXSPI_1PAD, 0x04),
+	},
+
+	/* Erase whole chip */
+	[NOR_CMD_LUT_SEQ_IDX_ERASECHIP] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, SPI_NOR_CMD_CE,
+				kFLEXSPI_Command_STOP, kFLEXSPI_1PAD, 0),
```
It's hard to tell which instructions actually changed after renaming and reordering them, and changing the whitespace too. Can you restore the original names and order so we can see more clearly where instructions changed?

pdgendt replied:
Sure, I'll try to get that cleaner. However can the enumeration of the instructions be changed? I followed this [NXP article](https://community.nxp.com/t5/i-MX-RT-Knowledge-Base/How-to-use-the-RT1064-on-chip-flash-as-NVM/ta-p/1123381) where it was explicitly mentioned to align the order to XIP. Or do you mean the order in which they are set in the lookup table itself?

MaureenHelm commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -283,12 +331,22 @@ static int flash_flexspi_nor_read(const struct device *dev, off_t offset,
 						     offset);
 
+	if (flash_flexspi_is_running_xip(data->controller)) {
+		key = irq_lock();
```
It shouldn't be necessary to lock interrupts during a read

MaureenHelm commented on dts/arm/nxp/nxp_rt1064.dtsi:
```diff
@@ -17,5 +17,7 @@
 		spi-max-frequency = <133000000>;
 		status = "okay";
 		jedec-id = [ef 40 16];
+		erase-block-size = <4096>;
+		write-block-size = <4>;
```
This doesn't look related to the commit message: `soc: nxp_imx: Disable IVT/DCD when building for mcuboot`

pdgendt replied:
True, but it is required if you want to sign the images for `mcuboot`.
Should I move this to another commit, or doesn't it belong to the PR at all?

Also the driver has this:
```C++
#define NOR_WRITE_SIZE  1
```

Which is hardcoded and doesn't match with what I added. Isn't the `dts` property preferred over the hardcoded value?

MaureenHelm replied:
Ah, I see. But why did you increase the write size to 4?

@galak how do you feel about this inheriting `soc-nv-flash.yaml` bindings?

MaureenHelm replied:
@galak ping

MaureenHelm reviewed (changes requested):
Thank you for the contribution!

pdgendt committed:
dts: add erase/write block sizes for imx rt1064 internal flash

This allows signing the binary for use with mcuboot

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

MaureenHelm commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -106,6 +120,24 @@ static int flash_flexspi_init(const struct device *dev)
 	flexspi_config_t flexspi_config;
 
+#ifdef CONFIG_XIP
+	struct flash_flexspi_data *data = dev->data;
+
+	if (config->base == DT_REG_ADDR(DT_NODELABEL(flexspi))) {
+		data->xip = IS_ENABLED(CONFIG_CODE_FLEXSPI);
+	} else if (config->base == DT_REG_ADDR(DT_NODELABEL(flexspi2))) {
+		data->xip = IS_ENABLED(CONFIG_CODE_FLEXSPI2);
+	} else {
+		LOG_ERR("Invalid base address");
+		return -EINVAL;
+	}
```
Can you fix the build warning:
```/home/maureen/zephyrproject/zephyr/drivers/flash/flash_mcux_flexspi.c: In function 'flash_flexspi_init':
/home/maureen/zephyrproject/zephyr/drivers/flash/flash_mcux_flexspi.c:126:19: warning: comparison between pointer and integer
  126 |  if (config->base == DT_REG_ADDR(DT_NODELABEL(flexspi))) {
      |                   ^~
/home/maureen/zephyrproject/zephyr/drivers/flash/flash_mcux_flexspi.c:128:26: warning: comparison between pointer and integer
  128 |  } else if (config->base == DT_REG_ADDR(DT_NODELABEL(flexspi2))) {
      |                          ^~
```

MaureenHelm reviewed (changes requested):
With some hacky changes to move the storage partition, I was able to confirm that `samples/subsys/fs/littlefs` can now work on the `mimxrt1064_evk` internal flash:
```diff
diff --git a/boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts b/boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts
index 13e0399c7d..03c86766c9 100644
--- a/boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts
+++ b/boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts
@@ -152,14 +152,30 @@ arduino_i2c: &lpi2c1 {};
                        #address-cells = <1>;
                        #size-cells = <1>;
 
-                       storage_partition: partition@0 {
-                               label = "storage";
+                       partition@0 {
                                reg = <0x00000000 DT_SIZE_M(8)>;
                        };
                };
        };
 };
 
+&flexspi2 {
+       status = "okay";
+};
+
+&w25q32jvwj0 {
+       partitions {
+               compatible = "fixed-partitions";
+               #address-cells = <1>;
+               #size-cells = <1>;
+
+               storage_partition: partition@14000 {
+                       label = "storage";
+                       reg = <0x00014000 DT_SIZE_K(32)>;
+               };
+       };
+};
+
 &lpuart1 {
        status = "okay";
        current-speed = <115200>;
```

mmahadevan108 commented on drivers/flash/Kconfig.mcux:
```diff
@@ -33,6 +33,40 @@ config FLASH_MCUX_FLEXSPI_NOR
+	string
+	default "ITCM" if FLASH_MCUX_FLEXSPI_XIP_MEM_ITCM
+	default "SRAM" if FLASH_MCUX_FLEXSPI_XIP_MEM_SRAM
+
```
Would we need to adjust MPU attributes if user chooses SRAM?

pdgendt replied:
This is something I don't have experience with, please advise on how this should be done.

Do I remove `SRAM` as a target option? I added it so the choice `Kconfig` would make sense
But we can change the MPU attributes if `FLASH_MCUX_FLEXSPI_XIP_MEM_SRAM` is selected.

mmahadevan108 replied:
I have added a PR:https://github.com/zephyrproject-rtos/zephyr/pull/34086 to fix this.

MaureenHelm commented:
Please rebase

pdgendt committed:
drivers: flash: mcux flexspi nor: Fix write if size > SPI_NOR_PAGE_SIZE

This fixes a bug in the write function of the MCUX FlexSPI flash driver
if the length of the data is larger than a single page.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: flash: mcux flexspi nor: copy data to RAM buffer on write

This feature prevents issues when trying to write data to the flash
device that is located on or addressed at the same flash device.

An example is the mcuboot logic that writes a magic number to
the secondary partition header.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

MaureenHelm commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -107,6 +115,11 @@ static int memc_flexspi_init(const struct device *dev)
 
+	/* we should not configure the device we are running on */
+	if (memc_flexspi_is_running_xip(dev)) {
+		return 0;
```
Can you add a LOG_DBG before the return?

pdgendt committed:
drivers: flash: Support i.MX FlexSPI NOR driver with XIP

This change allows writing to the flash while running in XIP mode,
and enables mcuboot or NVS settings to be used on i.MX RT socs.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

jfischer-no commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -19,6 +19,10 @@
 #define NOR_ERASE_VALUE	0xff
 
+#ifdef CONFIG_FLASH_MCUX_FLEXSPI_NOR_WRITE_BUFFER
+static uint8_t flash_flexspi_nor_write_buffer[SPI_NOR_PAGE_SIZE];
```
Would not shorter local name be clear enough here?
`static uint8_t nor_write_buf[SPI_NOR_PAGE_SIZE];`

jfischer-no commented on drivers/flash/Kconfig.mcux:
```diff
@@ -34,6 +34,18 @@ config FLASH_MCUX_FLEXSPI_NOR
+if FLASH_MCUX_FLEXSPI_NOR
+
+config FLASH_MCUX_FLEXSPI_NOR_WRITE_BUFFER
+	bool "MCUX FlexSPI NOR write RAM buffer"
```
`depends on FLASH_MCUX_FLEXSPI_NOR` is nicer than if/endif here.
